controlplane/kubeadm/internal/workload_cluster_conditions.go

```diff
@@ -298,76 +246,144 @@ func unwrapAll(err error) error {
+				Message: fmt.Sprintf("Failed to get etcd alarms: %s", unwrapAll(err)),
+			})
+		}
+		return nil, nil, errors.Wrapf(err, "failed to get etcd alarms")
 	}
 
 	return currentMembers, alarms, nil
 }
 
+// getNodeNamesSortedByLastKnownEtcdHealth return the list of nodes hosting an etcd member sorted by the last known etcd health.
+// Note: sorting by last known etcd health is a best effort operations; only nodes with a corresponding machine are considered.
+func getNodeNamesSortedByLastKnownEtcdHealth(nodes *corev1.NodeList, machines collections.Machines) []string {
+	// Get the list of nodes and the corresponding MachineEtcdMemberHealthyCondition
+	nodeNames := []string{}
+	nodeEtcdHealthyCondition := map[string]clusterv1.Condition{}
+
+	for _, node := range nodes.Items {
+		var machine *clusterv1.Machine
+		for _, m := range machines {
+			if m.Status.NodeRef != nil && m.Status.NodeRef.Name == node.Name {
+				machine = m
+				break
+			}
+		}
+		// Ignore nodes without a corresponding machine (this should never happen).
+		if machine == nil {
+			continue
+		}
+
+		nodeNames = append(nodeNames, node.Name)
+		if c := conditions.Get(machine, controlplanev1.MachineEtcdMemberHealthyCondition); c != nil {
+			nodeEtcdHealthyCondition[node.Name] = *c
+		}
+		nodeEtcdHealthyCondition[node.Name] = clusterv1.Condition{
+			Type:   controlplanev1.MachineEtcdMemberHealthyCondition,
+			Status: corev1.ConditionUnknown,
+		}
+	}
+
+	// Sort by nodes by last known etcd member health.
+	sort.Slice(nodeNames, func(i, j int) bool {
+		iCondition := nodeEtcdHealthyCondition[nodeNames[i]]
+		jCondition := nodeEtcdHealthyCondition[nodeNames[j]]
+
+		// Nodes with last known etcd healthy members goes first, because most likely we can connect to them again.
+		// NOTE: This isn't always true, it is a best effort assumption (e.g. kubelet might have issues preventing connection to an healthy member to be established).
+		if iCondition.Status == corev1.ConditionTrue && jCondition.Status != corev1.ConditionTrue {
+			return true
+		}
+		if iCondition.Status != corev1.ConditionTrue && jCondition.Status == corev1.ConditionTrue {
+			return false
+		}
+
+		// Note: we are not making assumption on the chances to connect when last known etcd health is FALSE and UNKNOWN.
+
+		// Otherwise pick randomly one of the nodes to avoid trying to connect always to the same nodes first.
+		return time.Now().UnixMilli()%2 == 0
+	})
+	return nodeNames
+}
+
 func compareMachinesAndMembers(controlPlane *ControlPlane, nodes *corev1.NodeList, members []*etcd.Member) (bool, []string) {
 	membersAndMachinesAreMatching := true
 	var kcpErrors []string
 
 	// If it failed to get members, consider the check failed in case there is at least a machine already provisioned
-	// (tolerate if we fail getting members when the cluster is provisioning the first machine).
+	// (tolerate if we fail getting members when the cluster is still creating or provisioning the first machine).
```

Comment:
Based on the added godoc I would have assume that we want to return true if `len(controlPlane.Machines) == 0`

Reply:
ACK, re reading this it makes sense to simplify